File: HowTheyWrite_android/app/src/main/java/com/kaiserpudding/howtheywrite/quiz/QuizViewModelFactory.java
package com.kaiserpudding.howtheywrite.quiz;

import android.app.Application;
import android.arch.lifecycle.ViewModel;
import android.arch.lifecycle.ViewModelProvider;
import android.support.annotation.NonNull;

public class QuizViewModelFactory implements ViewModelProvider.Factory {


  private Application application;
  private int lessonId;

  public QuizViewModelFactory(Application application, int lessonId) {
    this.application = application;
    this.lessonId = lessonId;
  }

  @NonNull
  @Override
  public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
    return (T) new QuizViewModel(application, lessonId);
  }
}

File: HowTheyWrite_android/app/src/main/java/com/kaiserpudding/howtheywrite/quiz/QuizViewModelFactory.kt
package com.kaiserpudding.howtheywrite.quiz

import android.app.Application
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider

/**
 * Factory to create a [QuizViewModel] with an additional parameter
 *
 * @property application
 * @property lessonId
 */
class QuizViewModelFactory : ViewModelProvider.Factory {

    private val application: Application
    private val lessonId: Long
    private val characterIds: LongArray?

    constructor(application: Application, lessonId: Long) {
        this.application = application
        this.lessonId = lessonId
        this.characterIds = null
    }

    constructor(application: Application, characterIds: LongArray) {
        this.application = application
        this.lessonId = -1
        this.characterIds = characterIds
    }

    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        return if (lessonId != -1L) QuizViewModel(application, lessonId) as T
        else QuizViewModel(application, characterIds!!) as T
    }
}
